man/man1/querypath.1
.TH "QUERYPATH" "1" "2021-04-12" "" "querypath"
.SH NAME
\fBquerypath\fR - a smarter cd command
.SH SYNOPSIS
.B querypath SUBCOMMAND [OPTIONS]
.SH DESCRIPTION
querypath is a smarter cd command for your terminal. It keeps track of the
directories you use most frequently, and uses a ranking algorithm to navigate
to the best match.
.SH USAGE
.nf
z foo              # cd into highest ranked directory matching foo
z foo bar          # cd into highest ranked directory matching foo and bar
z foo /            # cd into a subdirectory starting with foo
.sp
z ~/foo            # z also works like a regular cd command
z foo/             # cd into relative path
z ..               # cd one level up
z -                # cd into previous directory
.sp
zi foo             # cd with interactive selection (using fzf)
.sp
z foo<SPACE><TAB>  # show interactive completions (bash 4.4+/fish/zsh only)
.fi
.SH SUBCOMMANDS
.TP
\fBquerypath-add\fR(1)
Add a new directory to the database, or increment its rank.
.TP
\fBquerypath-import\fR(1)
Import entries from another application.
.TP
\fBquerypath-init\fR(1)
Generate shell configuration.
.TP
\fBquerypath-query\fR(1)
Search for a directory in the database.
.TP
\fBquerypath-remove\fR(1)
Remove a directory from the database.
.SH OPTIONS
.TP
.B -h, --help
Print help information.
.TP
.B -V, --version
Print version information.
.SH ENVIRONMENT VARIABLES
Environment variables can be used for configuration. They must be set before
\fBquerypath-init\fR(1) is called.
.TP
.B _QP_DATA_DIR
Specifies the directory in which the database is stored. The default value
varies across OSes:
.TS
tab(|);
l l.
    \fBOS|Path\fR
    \fBLinux/BSD\fR|T{
\fB$XDG_DATA_HOME\fR or \fB$HOME/.local/share\fR, eg.
\fB/home/alice/.local/share\fR
T}
    \fBmacOS\fR|T{
\fB$HOME/Library/Application Support\fR, eg.
\fB/Users/Alice/Library/Application Support\fR
T}
    \fBWindows\fR|T{
\fB%LOCALAPPDATA%\fR, eg. \fBC:\\Users\\Alice\\AppData\\Local\fR
T}
.TE
.TP
.B _QP_ECHO
When set to 1, \fBz\fR will print the matched directory before navigating to it.
.TP
.B _QP_EXCLUDE_DIRS
Prevents the specified directories from being added to the database. This is
provided as a list of globs, separated by OS-specific characters:
.TS
tab(|);
l l.
    \fBOS|Separator\fR
    \fBLinux/macOS/BSD\fR|T{
\fB:\fR, eg. \fB$HOME:$HOME/private/*\fR
T}
    \fBWindows\fR|\fB;\fR, eg. \fB$HOME;$HOME/private/*\fR
.TE
.sp
By default, this is set to \fB$HOME\fR. After setting this up, you might need
to use \fBquerypath-remove\fR(1) to remove any existing entries from the
database.
.TP
.B _QP_FZF_OPTS
Custom options to pass to \fBfzf\fR(1) during interactive selection. See the
manpage for the full list of options.
.TP
.B _QP_MAXAGE
Configures the aging algorithm, which limits the maximum number of entries in
the database. By default, this is set to 10000.
.TP
.B _QP_RESOLVE_SYMLINKS
When set to 1, \fBz\fR will resolve symlinks before adding directories to
the database.
.SH ALGORITHM
.TP
.B AGING
querypath uses a parameter called \fB_QP_MAXAGE\fR to limit the number of
entries in the database based on usage patterns. If the total \fBFRECENCY\fR of
the directories in the database exceeds this value, we divide each directory's
score by a factor \fBk\fR - such that the new total becomes ~90% of
\fB_QP_MAXAGE\fR. Thereafter, if the new score of any directory falls below
1, it is removed from the database.
.sp
Theoretically, the maximum number of directories in the database is
\fB4 * _QP_MAXAGE\fR, although it is lower in practice.
.TP
.B FRECENCY
Each directory in querypath is given a score, starting with 1 the first time
it is accessed. Every subsequent access increases the score by 1. When a
query is made, we calculate frecency based on the last time the directory was
accessed:
.TS
tab(|);
l l.
    \fBLast access time\fR|\fBFrecency\fR
    Within the last hour|score * 4
    Within the last day|score * 2
    Within the last week|score / 2
    Otherwise|score / 4
.TE
.SH REPORTING BUGS
For any issues, feature requests, or questions, please visit:
.sp
\fBhttps://github.com/dimarogiv/querypath/issues\fR
.SH AUTHOR
Ajeet D'Souza \fB<98ajeet@gmail.com>\fR


CHANGELOG.md
<!-- markdownlint-disable-file MD024 -->

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.9.8] - 2025-05-27

### Added

- Support for Tcsh.
- Added `--score` flag to `querypath add`.
- POSIX: add doctor to diagnose common issues.
- Nushell: add CLI completions.

### Changed

- Bash: querypath will now automatically `cd` when selecting Space-Tab completions.

### Fixed

- Bash: doctor now handles `PROMPT_COMMAND` being an array.
- Bash: doctor now handles Visual Studio Code's shell integration.
- Bash: completions now work with `ble.sh`.
- Nushell: stop ignoring symlinks when `cd`-ing into a directory.
- Fzf: updated minimum supported version to v0.51.0.
- PowerShell: avoid setting `$error` when defining `__querypath_hooked`.
- PowerShell: handle special characters in file paths when `cd`-ing into them.
- Database corruption issue when the filesystem is 100% full.

## [0.9.7] - 2025-02-10

### Added

- Nushell: support for 0.102.0.
- Bash / Zsh: add doctor to diagnose common issues.

### Fixed

- ksh: alias to regular POSIX implementation for better compatibility.

## [0.9.6] - 2024-09-19

### Fixed

- Fish: `builtin abbr` doesn't work on older versions.
- Zsh: make `__querypath_z_complete` available with `--no-cmd`.

## [0.9.5] - 2024-09-13

### Added

- Zsh: improved `cd` completions.
- Lazily delete excluded directories from the database.
- Fish: detect infinite loop when using `alias cd=z`.
- Installer: added flags for `--bin-dir`, `--man-dir`, `--arch`, and `--sudo`.
- Nushell: support for v0.94.0+.
- Bash/Fish/Zsh: support for `z -- dir` style queries.
- Fish: improved Space-Tab completions.
- Ksh: added support for the Korn shell.

### Changed

- fzf: removed `--select-1` from default options. The interactive selector will
  now open up even if there is only one match.
- Enforce that `$_QP_DATA_DIR` is an absolute path.

### Fixed

- Zsh: Space-Tab completion repeating output multiple times when matching single
  directory
- Fish / Nushell / PowerShell: handle queries that look like args (e.g. `z -x`).
- Elvish: `z -` now works as expected.
- Fish: generated shell code avoids using aliased builtins.
- Fish: `cd` command is now copied directly from
  `$__fish_data_dir/functions/cd.fish`. This should minimize the chances of an
  infinite loop when aliasing `cd=z`.
- Symlinks not getting added to the database when `$_QP_RESOLVE_SYMLINKS=0`.
- Symlinked database files getting replaced instead of the actual files.

## [0.9.4] - 2024-02-21

### Changed

- Zsh: improved Space-Tab completions.

## [0.9.3] - 2024-02-13

### Added

- Nushell: support for v0.89.0.

## [0.9.2] - 2023-08-04

### Added

- Short option `-a` for `querypath query --all`.

### Fixed

- PowerShell: use `global` scope for variables / functions.

## [0.9.1] - 2023-05-07

### Added

- Fish/Zsh: aliases on `__querypath_z` will now use completions.
- Nushell: support for v0.78.0.
- Fish: plugin now works on older versions.
- PowerShell: warn when PowerShell version is too old for `z -` and `z +`.
- PowerShell: support for PWD hooks on all versions.

### Fixed

- Fish: not providing `cd` completions when there is a space in the path.
- Bash/Fish/Zsh: providing `z` completions when the last argument starts with `z!`.
- Bash/Fish/Zsh: attempting to `cd` when the last argument is `z!`.

## [0.9.0] - 2023-01-08

### Added

- `edit` subcommand to adjust the scores of entries.

### Fixed

- Zsh: completions clashing with `zsh-autocomplete`.
- Fzf: 'invalid option' on macOS.
- PowerShell: handle UTF-8 encoding correctly.
- Zsh: don't hide output from `chpwd` hooks.
- Nushell: upgrade minimum supported version to v0.73.0.
- Zsh: fix extra space in interactive completions when no match is found.
- Fzf: various improvements, upgrade minimum supported version to v0.33.0.
- Nushell: accidental redefinition of hooks when initialized twice.

### Removed

- `remove -i` subcommand: use `edit` instead.

## [0.8.3] - 2022-09-02

### Added

- Nushell: support for `z -`.
- Nushell: support for PWD hooks.

### Changed

- Fish: change fuzzy completion prefix to `z!`.
- Zsh: allow `z` to navigate dirstack via `+n` and `-n`.
- Fzf: improved preview window.

### Fixed

- Bash: double forward slash in completions.

## [0.8.2] - 2022-06-26

### Changed

- Fzf: show preview window below results.

### Fixed

- Bash/Fish/POSIX/Zsh: paths on Cygwin.
- Fish: completions not working on certain systems.
- Bash: completions not escaping spaces correctly.

## [0.8.1] - 2021-04-23

### Changed

- Manpages: moved to `man/man1/*.1`.
- Replace `--no-aliases` with `--no-cmd`.
- Elvish: upgrade minimum supported version to v0.18.0.
- Nushell: upgrade minimum supported version to v0.61.0.

### Fixed

- Bash/Zsh: rename `_z` completion function to avoid conflicts with other shell
  plugins.
- Fzf: added `--keep-right` option by default, upgrade minimum supported version
  to v0.21.0.
- Bash: only enable completions on 4.4+.
- Fzf: bypass `ls` alias in preview window.
- Retain ownership of database file.
- `querypath query --interactive` should not conflict with `--score`.

## [0.8.0] - 2021-12-25

### Added

- Zsh: completions for `z` command.

### Changed

- Fzf: better default options.
- Fish: interactive completions are only triggered when the last argument is
  empty.
- PowerShell: installation instructions.

### Fixed

- PowerShell: use global scope for aliases.
- Zsh: fix errors with `set -eu`.
- Fzf: handle early selection.
- PowerShell: correctly handle escape characters in paths.
- Parse error on Cygwin/MSYS due to CRLF line endings.
- Fzf: handle spaces correctly in preview window.
- Bash: avoid initializing completions on older versions.
- Fzf: avoid launching binary from current directory on Windows.

## [0.7.9] - 2021-11-02

### Changed


contrib/completions/querypath.ts
const completion: Fig.Spec = {
  name: "querypath",
  description: "A smarter cd command for your terminal",
  subcommands: [
    {
      name: "add",
      description: "Add a new directory or increment its rank",
      options: [
        {
          name: ["-s", "--score"],
          description: "The rank to increment the entry if it exists or initialize it with if it doesn't",
          isRepeatable: true,
          args: {
            name: "score",
            isOptional: true,
          },
        },
        {
          name: ["-h", "--help"],
          description: "Print help",
        },
        {
          name: ["-V", "--version"],
          description: "Print version",
        },
      ],
      args: {
        name: "paths",
        isVariadic: true,
        template: "folders",
      },
    },
    {
      name: "edit",
      description: "Edit the database",
      subcommands: [
        {
          name: "decrement",
          hidden: true,
          options: [
            {
              name: ["-h", "--help"],
              description: "Print help",
            },
            {
              name: ["-V", "--version"],
              description: "Print version",
            },
          ],
          args: {
            name: "path",
          },
        },
        {
          name: "delete",
          hidden: true,
          options: [
            {
              name: ["-h", "--help"],
              description: "Print help",
            },
            {
              name: ["-V", "--version"],
              description: "Print version",
            },
          ],
          args: {
            name: "path",
          },
        },
        {
          name: "increment",
          hidden: true,
          options: [
            {
              name: ["-h", "--help"],
              description: "Print help",
            },
            {
              name: ["-V", "--version"],
              description: "Print version",
            },
          ],
          args: {
            name: "path",
          },
        },
        {
          name: "reload",
          hidden: true,
          options: [
            {
              name: ["-h", "--help"],
              description: "Print help",
            },
            {
              name: ["-V", "--version"],
              description: "Print version",
            },
          ],
        },
      ],
      options: [
        {
          name: ["-h", "--help"],
          description: "Print help",
        },
        {
          name: ["-V", "--version"],
          description: "Print version",
        },
      ],
    },
    {
      name: "import",
      description: "Import entries from another application",
      options: [
        {
          name: "--from",
          description: "Application to import from",
          isRepeatable: true,
          args: {
            name: "from",
            suggestions: [
              "autojump",
              "z",
            ],
          },
        },
        {
          name: "--merge",
          description: "Merge into existing database",
        },
        {
          name: ["-h", "--help"],
          description: "Print help",
        },
        {
          name: ["-V", "--version"],
          description: "Print version",
        },
      ],
      args: {
        name: "path",
        template: "filepaths",
      },
    },
    {
      name: "init",
      description: "Generate shell configuration",
      options: [
        {
          name: "--cmd",
          description: "Changes the prefix of the `z` and `zi` commands",
          isRepeatable: true,
          args: {
            name: "cmd",
            isOptional: true,
          },
        },
        {
          name: "--hook",
          description: "Changes how often querypath increments a directory's score",
          isRepeatable: true,
          args: {
            name: "hook",
            isOptional: true,
            suggestions: [
              "none",
              "prompt",
              "pwd",
            ],
          },
        },
        {
          name: "--no-cmd",
          description: "Prevents querypath from defining the `z` and `zi` commands",
        },
        {
          name: ["-h", "--help"],
          description: "Print help",
        },
        {
          name: ["-V", "--version"],
          description: "Print version",
        },
      ],
      args: {
        name: "shell",
        suggestions: [
          "bash",
          "elvish",
          "fish",
          "nushell",
          "posix",
          "powershell",
          "tcsh",
          "xonsh",
          "zsh",
        ],
      },
    },
    {
      name: "query",
      description: "Search for a directory in the database",
      options: [
        {
          name: "--exclude",
          description: "Exclude the current directory",
          isRepeatable: true,
          args: {
            name: "exclude",
            isOptional: true,
            template: "folders",
          },
        },
        {
          name: ["-a", "--all"],
          description: "Show unavailable directories",
        },
        {
          name: ["-i", "--interactive"],
          description: "Use interactive selection",
          exclusiveOn: [
            "-l",
            "--list",
          ],
        },
        {
          name: ["-l", "--list"],
          description: "List all matching directories",
          exclusiveOn: [
            "-i",
            "--interactive",
          ],
        },
        {
          name: ["-s", "--score"],
          description: "Print score with results",
        },
        {
          name: ["-h", "--help"],
          description: "Print help",
        },
        {
          name: ["-V", "--version"],
          description: "Print version",
        },
      ],
      args: {
        name: "keywords",
        isVariadic: true,
        isOptional: true,
      },
    },
    {
      name: "remove",
      description: "Remove a directory from the database",
      options: [
        {
          name: ["-h", "--help"],
          description: "Print help",
        },
        {
          name: ["-V", "--version"],
          description: "Print version",
        },
      ],
      args: {
        name: "paths",
        isVariadic: true,
        isOptional: true,
        template: "folders",
      },
    },
  ],
  options: [
    {
      name: ["-h", "--help"],
      description: "Print help",
    },
    {
      name: ["-V", "--version"],
      description: "Print version",
    },
  ],
};

export default completion;
